doc: various developer notes updates

jonatack opened the pull request:
A few updates noticed while working on a lifetimebound section.

- Remove obsolete BDB ENABLE_WALLET section (only one file, src/wallet/bdb.h, still has a `db_cxx.h` BDB header)
- Prefer Python for scripts in developer notes (and a few miscellaneous touch-ups)
- In the code style section, add a link to the internal interface exception so that people are aware of it
- Add a "C++ functions and methods" section
- Add a Clang `lifetimebound` attribute section

jonatack committed:
Remove obsolete BDB ENABLE_WALLET section in developer notes

jonatack committed:
Prefer Python for scripts in developer notes

along with a few miscellaneous touch-ups.

jonatack committed:
Link in developer notes style to internal interface exception

laanwj commented on doc/developer-notes.md:
```diff
@@ -1390,7 +1391,7 @@ communication:
 - For consistency and friendliness to code generation tools, interface method
-  input and inout parameters should be ordered first and output parameters
+  input and in-out parameters should be ordered first and output parameters
   should come last.
```
Maybe we can add something that we prefer returning values directly to "output parameters". This was written before the days of `std::optional` etc.

jonatack replied:
Good idea, added 968a67b9be5262228b47184b9821fcab as a first attempt.

jonatack committed:
Add C++ functions and methods section to developer notes

Credit for some parts to the Google C++ Style Guide "Inputs and Outputs"
section at https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

Co-authored-by: laanwj <126646+laanwj@users.noreply.github.com>

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/25092/commits/91d6d4a2337719cce48532dd614bb345be66fbbf

laanwj commented on doc/developer-notes.md:
```diff
@@ -138,6 +141,16 @@ public:
+--------------------
+
+- Input and in-out parameters should be ordered first and output parameters
+  should come last.
```
```
- *Rationale*: API consistency.
- ```

jonatack replied:
Added (thanks!)

laanwj commented on doc/developer-notes.md:
```diff
@@ -138,6 +141,16 @@ public:
+  should come last.
+
+- Prefer returning values directly (using `std::optional`, where helpful) to
+  using output parameters.
```
```
- *Rationale*: Less error-prone (no need for assumptions what output is initialized to on failure). Easier to read.
```

jonatack replied:
Added (thanks!) along with some ideas from https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs.

laanwj commented on doc/developer-notes.md:
```diff
@@ -138,6 +141,22 @@ public:
+- *Rationale*: API consistency.
+
+- Prefer returning values directly to using in-out or output parameters. Use
+  `std::optional` where helpful for returning values (as well as for optional
```
Good point. I think using `std::optional` *for optional inputs* (instead some magic default value, if there is no real default) has enough merit to be called out separately.

jonatack replied:
> has enough merit to be called out separately.

Suggesting splitting it out to a separate bullet point?

laanwj replied:
I do think that's clearer.

jonatack replied:
Done.

jonatack committed:
Add clang lifetimebound section to developer notes

jarolrod reviewed (commented):
code review ACK https://github.com/bitcoin/bitcoin/commit/654284209f30fd309c326a527cda1d2d7385cee8

Nice improvements to the dev docs 👍
